### PR TITLE
gl2d data-referenced annotations

### PR DESCRIPTION
**Pull request by etpinard**
set to :heavy_check_mark: one item on https://github.com/plotly/plotly.js/issues/130

gifrecord_2017-01-13_171012

TODO:

- [x] add some jasmine test (and maybe clean up our gl2d test suites along the way)

### CODE REVIEW

File: src/plots/gl2d/scene2d.js [hunk 1]
```diff
@@ -286,10 +292,11 @@ proto.updateFx = function(options) {
 
     fullLayout.dragmode = options.dragmode;
     fullLayout.hovermode = options.hovermode;
-};
 
-var relayoutCallback = function(scene) {
+    this.graphDiv._fullLayout = fullLayout;
```
Comment (alexcjohnson) [inline src/plots/gl2d/scene2d.js]: this seems a bit scary - why is this necessary? I'm just worried that if `gd._fullLayout` and `scene.fullLayout` have diverged, there could somehow be changes in `gd._fullLayout` that this will throw away.

Comment (etpinard) [inline src/plots/gl2d/scene2d.js]: I believe the problem is in the [`doModebar`](https://github.com/plotly/plotly.js/blob/5e4bb1a49f93e59e682a142099923dad3a21f861/src/plot_api/subroutines.js#L293-L319) relayout shortcut. 

I'll double-check why the `_fullLayout` gets lost. Thanks for the headsup!

Comment (alexcjohnson) [inline src/plots/gl2d/scene2d.js]: would it be an option to not use `this.fullLayout` at all, but only to use `this.graphDiv._fullLayout`?

Comment (etpinard) [inline src/plots/gl2d/scene2d.js]: That's an idea.

So, the `_fullLayout`  gets lost in `Plots.supplyDefaults`. We could add a block handling refs on scene objects similar to this [one](https://github.com/plotly/plotly.js/blob/9b7c18f5f576f103b65889574d25ad9924d4d7c1/src/plots/plots.js#L494-L499) for cartesian axes. But I think we could do better.

The reason the lost of reference only affects `dragmode` relayouts is because all other relayout calls pass through `scene2d.plot` which update the `fullLayout` ref [here](https://github.com/plotly/plotly.js/blob/9b7c18f5f576f103b65889574d25ad9924d4d7c1/src/plots/gl2d/scene2d.js#L353) on every pass.

Comment (alexcjohnson) [inline src/plots/gl2d/scene2d.js]: alright, well I'm not entirely sure what would happen, but the test would be something like:
- do a relayout that causes a new `gd._fullLayout` to be constructed by `supplyDefaults` but DOESN'T result in a new `scene2d.plot` (like what? change legend background color? change something cosmetic on an axis of a different subplot?)
- trigger this code by changing dragmode
- do something else that causes a full plot redraw (maybe just `Plotly.redraw(gd)`)
- does the initial relayout get reverted?

Comment (etpinard) [inline src/plots/gl2d/scene2d.js]: Here's the _cleanest_ way (I think) of updating `gd._fullLayout._plots.xy._scene2d.fullLayout`

https://github.com/plotly/plotly.js/commit/b50b7e99bb8e720240fae7c4ae0e8e3a1bb8a0ff

Comment (alexcjohnson) [inline src/plots/gl2d/scene2d.js]: yeah, that seems better, thanks.

Comment (etpinard) [inline src/plots/gl2d/scene2d.js]: full solution in https://github.com/plotly/plotly.js/pull/1319
